### PR TITLE
Adjust hue accepts torch tensor

### PR DESCRIPTION
**Pull request by vikramtankasali**
Add functionality to adjust hue for torch tensor type image.

### CODE REVIEW

File: torchvision/transforms/functional_tensor.py [hunk 1]
```diff
@@ -236,3 +276,51 @@ def _blend(img1, img2, ratio):
...
+
+    maxc, _ = torch.max(img, dim=0)
+    minc, _ = torch.min(img, dim=0)
+    uv = maxc
+
+    cr = maxc.to(dtype=torch.float32) - minc.to(dtype=torch.float32)
+    s = cr / maxc.to(dtype=torch.float32)
+    rc = (maxc-r).to(dtype=torch.float32) / cr
+    gc = (maxc-g).to(dtype=torch.float32) / cr
+    bc = (maxc-b).to(dtype=torch.float32) / cr
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:292]: If we move the casting to float in the beginning of the function, we can avoid casting individually each tensor.
Something like

```suggestion
    orig_dtype = img.dtype
    img = img.to(dtype=torch.float32)
    r, g, b = img.unbind(0)

    maxc, _ = torch.max(img, dim=0)
    minc, _ = torch.min(img, dim=0)
    uv = maxc

    cr = maxc - minc
    s = cr / maxc
    rc = (maxc-r) / cr
    gc = (maxc-g) / cr
    bc = (maxc-b) / cr
```

File: torchvision/transforms/functional_tensor.py [hunk 2]
```diff
@@ -236,3 +276,51 @@ def _blend(img1, img2, ratio):
...
+
+    #  torch.fmod and  math.fmod have different precision.
+    h = torch.fmod((h / 6.0 + 1.0), 1.0)
+
+    if img.dtype == torch.uint8:
+        uh = torch.clamp((h*255.0).to(dtype=torch.int32), 0, 255).to(dtype=torch.uint8)
+        us = torch.clamp((s*255.0).to(dtype=torch.int32), 0, 255).to(dtype=torch.uint8)
+    else:
+        uh = torch.clamp(h * 255.0, 0.0, 255.0)
+        us = torch.clamp(s * 255.0, 0.0, 255.0)
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:308]: Given that this function normalizes by the max value, it might make sense to expect the tensor to be fed to it to be a float tensor, and to always return float tensors in 0-1 for this helper function, and then use the (yet not merged) `convert_image_dtype` https://github.com/pytorch/vision/pull/2078 to cast from float to the original dtype.

File: torchvision/transforms/functional_tensor.py [hunk 3]
```diff
@@ -236,3 +276,51 @@ def _blend(img1, img2, ratio):
...
+    cr = maxc.to(dtype=torch.float32) - minc.to(dtype=torch.float32)
+    s = cr / maxc.to(dtype=torch.float32)
+    rc = (maxc-r).to(dtype=torch.float32) / cr
+    gc = (maxc-g).to(dtype=torch.float32) / cr
+    bc = (maxc-b).to(dtype=torch.float32) / cr
+
+    s = (maxc != minc) * s
+    hr = (maxc == r) * (bc - gc)
+    hg = ((maxc == g) & (maxc != r)) * (2.0 + rc - bc)
+    hb = ((maxc != g) & (maxc != r)) * (4.0 + gc - rc)
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:297]: I think if you keep the indices from
```python
maxc, max_idxs = torch.max(img, dim=0)
```
then you can avoid these set of equalities to find where the max came from. This part could even be batched and vectorized
```python
max_mask = max_idxs[..., None] == torch.tensor([0, 1, 2])
temp1 = (bc - gc)
temp2 = (2.0 + rc - bc)
temp3 = (4.0 + gc - rc)
temp = torch.stack([temp1, temp2, temp3], dim=-1)
h = (max_mask * temp).sum(-1)
```

Although having 2 additions might be more efficient than calling `sum`

File: torchvision/transforms/functional_tensor.py [hunk 4]
```diff
@@ -236,3 +276,51 @@ def _blend(img1, img2, ratio):
...
+        uh = torch.clamp(h * 255.0, 0.0, 255.0)
+        us = torch.clamp(s * 255.0, 0.0, 255.0)
+
+    return torch.stack((uh, us, uv))
+
+def _hsv2rgb(img):
+    h, s, v = img[0], img[1], img[2]
+    i = torch.floor(h.to(dtype=torch.float32) * 6.0 / 255.0).to(dtype=torch.int32)
+    f = (h.to(dtype=torch.float32) * 6.0 / 255.0) - i.to(dtype=torch.float32)
+    fs = s.to(dtype=torch.float32) / 255.0
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:316]: Similarly here, I think it might be preferable to assume that the inputs are float in 0-1 and in the exact same format as what the `rgb2hsv` returns, so we can avoid performing those divisions in the function.

File: torchvision/transforms/functional_tensor.py [hunk 5]
```diff
@@ -236,3 +276,51 @@ def _blend(img1, img2, ratio):
...
+
+    p = torch.clamp(torch.round(v.to(dtype=torch.float32) * (1.0 - fs)), 0, 255).to(dtype=img.dtype)
+    q = torch.clamp(torch.round(v.to(dtype=torch.float32) * (1.0 - fs * f)), 0, 255).to(dtype=img.dtype)
+    t = torch.clamp(torch.round(v.to(dtype=torch.float32) * (1.0 - fs * (1.0 - f))), 0, 255).to(dtype=img.dtype)
+    i = i % 6
+
+    r = (i == 0) * v + (i == 1) * q +  (i == 2) * p + (i == 3) * p + (i == 4) * t + (i == 5) * v
+    g = (i == 0) * t + (i == 1) * v +  (i == 2) * v + (i == 3) * q + (i == 4) * p + (i == 5) * p
+    b = (i == 0) * p + (i == 1) * p +  (i == 2) * t + (i == 3) * v + (i == 4) * v + (i == 5) * q
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:325]: This is fine, but one could probably batch the equalities as well, something like
```python
mask = i[..., None] == torch.arange(6)
```
which uses broadcasting.
The dimension to broadcast is TBD, could be either the first or the last dimension (maybe the first dimension is better here?)

File: test/test_functional_tensor.py [hunk 6]
```diff
@@ -56,6 +56,38 @@ def test_crop(self):
...
+
+                ft_img = ft(img, factor)
+                sft_img = sft(img, factor)
+
+                img_pil = transforms.ToPILImage()(img)
+                f_img_pil = f(img_pil, factor)
+                f_img = np.array(f_img_pil, dtype=np.float)
+
+                diff = (ft_img.to(dtype=torch.float) - f_img.astype(float)).norm().numpy()
+                rel_err =  diff / np.linalg.norm(f_img.astype(float))
```
Comment (fmassa) [inline test/test_functional_tensor.py:83]: Instead of returning the relative error by dividing the norm, which is a function of the number of elements, can you return as well the max absolute difference between the predicted value and the expected value?
```python
diff = (ft_img - f_img).abs().max()
```
?
In practice, I've found it to be useful to account of rounding errors, and it is informative of the max error that you are having. Ideally it should be 0, but due to numerical differences it could be slightly more as well.

File: torchvision/transforms/functional_tensor.py (hunk 3 above)
Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py:297]: Also i have to ensure I dont add twice if the img has duplicate values. I dont see how the above code avoids that from happening. Please let me know.

File: test/test_functional_tensor.py (hunk 6 above)
Comment (vikramtankasali) [reply test/test_functional_tensor.py:83]: I tried this "diff". However i have seen that torch.fmod (used in rgb2hsv) and fmod have different precision. This results in ft_img and f_img differing by 1 (or 255). This creates issues with the testing.

File: torchvision/transforms/functional_tensor.py (hunk 1 above)
Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py:292]: Done.

Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py:292]: Done

File: torchvision/transforms/functional_tensor.py (hunk 3 above)
Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py:297]: The main problem I was facing with this was when img has duplicate values and the max occurs at two or more locations. In that case there is an order that needs to be followed. For instance maxc == r > maxc ==g > maxc = b. And we need to add only one of the rc, gc, or bc to h.  I dont understand how we can enforce this order and mutually exclusive thing from your suggestion. Please let me know.

File: test/test_functional_tensor.py (hunk 6 above)
Comment (vikramtankasali) [reply test/test_functional_tensor.py:83]: The main problem is that imgs are uint8 and a small difference (1 or 255) gets blown up. torch.fmod and math.fmod have different precision and when I was testing I saw rgb2hsv returning h values which differ from baseline by 1 or 2. This causes issues with the testing. Please let me know.

File: torchvision/transforms/functional_tensor.py (hunk 3 above)
Comment (fmassa) [reply torchvision/transforms/functional_tensor.py:299]: Sounds good. There could be ways of working around this issue, but it would probably make the code less readable.

File: test/test_functional_tensor.py (hunk 6 above)
Comment (fmassa) [reply test/test_functional_tensor.py:83]: The difference of 255 happens even if you cast to float before the subtraction in the test?
Also, is the difference due mainly to fmod or do you think it could come from somewhere else?

Comment (vikramtankasali) [reply test/test_functional_tensor.py:83]: No.  Before cast.  rgb2hsv fmod was the only place where i found this issue happens.

Comment (fmassa) [reply test/test_functional_tensor.py:83]: So, `torch.fmod` should be similar to `fmod` in C. Would using `torch.remainder` fix the issue?

File: torchvision/transforms/functional_tensor.py (hunk 2 above)
Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py]: This function normalizes h by maxc-minc, s by maxc and v is not normalized. I dont completelt get your point here. I can change this function to accept [0, 1]. But I am wondering if we keep it the same as in PIL will it be better?

File: torchvision/transforms/functional_tensor.py (hunk 4 above)
Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py:319]: After rgb2hsv I am adding the hue factor which needs to be done in uint8 to avoid round of issues (We can mod there anyway i think). But again should we keep it similar to PIL implementation? I have now simplified the casts so that we dont perform cast on h, s, v, more than once.

File: test/test_functional_tensor.py (hunk 6 above)
Comment (vikramtankasali) [reply test/test_functional_tensor.py:83]: I tried torch.remainder. I tried debugging  math.fmod vs torch.fmod there is a small difference which when multiplied by 255.0 and subsequently rounded is causing the discrepancy.

File: torchvision/transforms/functional_tensor.py (hunk 2 above)
Comment (fmassa) [reply torchvision/transforms/functional_tensor.py]: So, here is what I meant (and I missed that `v` was not invariant, sorry about that):
- both `h` and `s` are invariant by scaling, so `_rgb2hsv(img)[0:2] == _rgb2hsv(img * 255)[0:2]` (minus floating point errors)
- `v` is equivariant by scaling, so `_rgb2hsv(img)[2] == _rgb2hsv(img * 255)[2] / 255`

This means that the range of the input doesn't matter for `h` and `s`, and the range for `v` is just scaled according to the max value.

This means that, for simplicity, we can remove the `* 255` in this part of the function (which is specific to `uint8` representation), and instead scale the whole output tensor by the `convert_image_dtype` (or equivalent) in the main function, as it would make things simpler (and could potentially support other image type representations)

File: torchvision/transforms/functional_tensor.py [hunk 7]
```diff
@@ -236,3 +280,48 @@ def _blend(img1, img2, ratio):
...
+
+    mask = i[..., None] == torch.arange(6)
+    mask = mask.transpose(1, 2).transpose(0,1)
+
+    a1=torch.stack((v, q, p, p, t, v))
+    a2=torch.stack((t, v, v, q, p, p))
+    a3=torch.stack((p, p, t, v, v, q))
+    a4 = torch.stack((a1, a2, a3))
+
+    return torch.einsum("ijk, xijk -> xjk", mask.to(dtype=torch.float32), a4)
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:327]: instead of forcing the `mask` to be `float32` here, can you instead do (in like 319)
```
mask = mask.to(img)
```
so that it works for double tensors as well, as it will take the dtype of the `img` (and device)

File: torchvision/transforms/functional_tensor.py [hunk 8]
```diff
@@ -119,6 +119,50 @@ def adjust_contrast(img, contrast_factor):
...
+
+    orig_dtype = img.dtype
+    if img.dtype == torch.uint8:
+        img = img.to(dtype=torch.float32) / 255.0
+
+    img = _rgb2hsv(img)
+    h, s, v = img.unbind(0)
+    h += hue_factor
+    img = torch.stack((h, s, v))
+    img_hue_adj = _hsv2rgb(img).transpose(0, 1).transpose(1, 2)
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:161]: Can we return instead a tensor of shape `[C, H, W]`? This is the standard convention in PyTorch.

File: test/test_functional_tensor.py [hunk 9]
```diff
@@ -56,6 +57,46 @@ def test_crop(self):
...
+            colorsys_img = torch.tensor(rgb, dtype=torch.float32)
+            max_diff = (ft_img- colorsys_img).abs().max()
+            self.assertLess(max_diff, 1e-5)
+
+    def test_rgb2hsv(self):
+         shape = (3, 150, 100)
+         for _ in range(20):
+            img = torch.rand(*shape, dtype=torch.float)
+
+            ft_hsv_img = F_t._rgb2hsv(img).transpose(0, 1).transpose(1, 2).reshape(np.prod(shape[1:]), shape[0])
```
Comment (fmassa) [inline test/test_functional_tensor.py:84]: Same nit here about using `permute` and `flatten`

File: torchvision/transforms/functional_tensor.py (hunk 2 above)
Comment (vikramtankasali) [reply torchvision/transforms/functional_tensor.py]: Done.
